blocks/bricks/__init__.py

```diff
@@ -1,22 +1,324 @@
+    the :meth:`__get__` method) automatically instantiates a
+    :class:`BoundApplication` class and returns this. This bound
+    application class can be used to store application information
+    particular to a brick instance. Any attributes unknown to the bounded
+    application are automatically routed to the application that
+    instantiated it.
+
+    """
+    call_stack = []
+
+    def __init__(self, application):
+        self.application = application
+        self.delegate_function = None
+        self.properties = {}
+        self.instances = {}
+
+    def property(self, name):
+        """Decorator to make application properties.
+
+        Parameters
+        ----------
+        name : str
+            The name the property should take.
+
+        Examples
+        --------
+        >>> class Foo(Brick):
+        ...     @application
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.property('inputs')
+        ...     def apply_inputs(self):
+        ...         return ['foo', 'bar']
+        >>> foo = Foo()
+        >>> foo.apply.inputs
+        ['foo', 'bar']
+
+        """
+        if not isinstance(name, six.string_types):
+            raise ValueError
+
+        def wrap_property(property_):
+            self.properties[name] = property_
+            return property_
+        return wrap_property
+
+    def delegate(self, f):
+        """Decorator to assign a delegate application.
+
+        An application method can assign a delegate application. Whenever
+        an attribute is not available, it will be requested from the
+        delegate instead.
+
+        Examples
+        --------
+        >>> class Foo(Brick):
+        ...     @application(outputs=['baz'])
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.property('inputs')
+        ...     def apply_inputs(self):
+        ...         return ['foo', 'bar']
+        >>> class Bar(Brick):
+        ...     def __init__(self, foo):
+        ...         self.foo = foo
+        ...
+        ...     @application(outputs=['foo'])
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.delegate
+        ...     def apply_delegate(self):
+        ...         return self.foo.apply
+        >>> foo = Foo()
+        >>> bar = Bar(foo)
+        >>> bar.apply.outputs
+        ['foo']
+        >>> bar.apply.inputs
+        ['foo', 'bar']
+
+        """
+        self.delegate_function = f
+        return f
+
+    def __get__(self, instance, owner):
+        """Instantiate :class:`BoundedApplication` for each :class:`Brick`."""
+        if instance is None:
+            return self
+        elif instance in self.instances:
+            return self.instances[instance]
+        else:
+            bounded_application = BoundApplication(self, instance)
```

Comment:
It's "bound" when referring to being confined to something. "Bounded" is
when you're talking about running. You'll see that Python methods are also
bound, not bounded.
On Jan 22, 2015 3:13 AM, "Dmitry Bogdanov" <notifications@github.com> wrote:

> In blocks/bricks/__init__.py
> <https://github.com/bartvm/blocks/pull/127#discussion_r23360369>:
>
> > +        ['foo']
> > +        >>> bar.apply.inputs
> > +        ['foo', 'bar']
> > +
> > +        """
> > +        self.delegate_function = f
> > +        return f
> > +
> > +    def __get__(self, instance, owner):
> > +        """Instantiate :class:`BoundedApplication` for each :class:`Brick`."""
> > +        if instance is None:
> > +            return self
> > +        elif instance in self.instances:
> > +            return self.instances[instance]
> > +        else:
> > +            bounded_application = BoundApplication(self, instance)
>
> Linguistic nitpicking: I guess you wanted to use the third form of the
> verb bind, which is bound, not bounded.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bartvm/blocks/pull/127/files#r23360369>.
>